docs/proposals/separate-config.md

```diff
@@ -0,0 +1,32 @@
+We should separate the config for our separate processes.  We have
```

Comment:
I almost, but don't quite agree with the first statement.  I think there are three broad classes of config:

1. core config that drives the behavior of the system, that should be specified in exactly one place (example - feature enablement)
2. config that most people should never set but should be automatically configured because they'd only screw it up
3. specific config that guides the automatic config above

If you're saying that we have a top level config that drives the operator which drives the per process config, we're closer.

The problem with the per process model is that it is actively encouraging lumping together 1, 2, and 3.  Using per-process config for 2 is totally acceptable.  However, I am very against having the config for 3 and 1 be per process or not clearly designed to limit total choice and expose the consistent choices.  Tectonic's approach here is superior both to our current "giant bucket approach" as well as the pure per process model.  Our process model should not be user facing - it is an implementation detail that should make sense on its own, but no administrator should be mucking with it.

I do *not* want users editing the registry config directly to set GCS bucket rules (they should be setting a config map setting for "registry: gcs"), or the admission plugin list to change image policy.  I *do* want users editing image policy rules via a config map directly.

Separate from the top level feature enablement , a few things that aren't clearly per process

1. feature gates (which is like feature enablement, but simpler - I don't necessarily think those should be coupled)
1. the default router subdomain (although we could expose it another way)

Reply:
The choice of how a process accepts it's config is distinct from how that config is built.   Components with config dependencies largely rely on the status of that config.  Downstream pieces don't get to choose it, they have to consume it.  An example of this is the webconsole consuming the logging route URL. The webconsole doesn't get to choose that config.  It accepts the status (reality) from the logging config and reacts.  Wiring as an operator setting config, you'd modify the configmap with your config and kick the webconsole pods.

The same applies here.  When components rely on the status of other configs, the operator configuring the downstream components needs to detect that status and react, but how the process gets its own config is distinct from how it gets wired.

That perspective means that case 1 is twiddled at a higher level, but each piece still receives config independently.  If you choose to go in the other direction, you'll end up with a larger and larger mono-config that fails to scale out for additional components.